### PR TITLE
APP-2888: using AppValidationErrors to show errors with modules

### PR DESCRIPTION
**Pull request by RoxyFarhad**
This goes hand in hand with this change: https://github.com/viamrobotics/app/pull/2672
(must wait for API + then APP change before we merge)

### CODE REVIEW

File: go.sum
Comment (RoxyFarhad) [inline go.sum:1585]: once app is unblocked to bump api, I will re-pull the latest api change into RDK

File: module/modmanager/manager.go
Comment (RoxyFarhad) [inline module/modmanager/manager.go:703]: Changed this as part of another small ticket in rdk: https://viam.atlassian.net/browse/APP-2700?atlOrigin=eyJpIjoiYTNlZDUyMzllYzc3NDU0YjgxYTUxNjdiNjc5Y2I5YTYiLCJwIjoiaiJ9

File: module/modmanager/manager.go
Comment (RoxyFarhad) [inline module/modmanager/manager.go:717]: Changed this as part of another small ticket in rdk: https://viam.atlassian.net/browse/APP-2700?atlOrigin=eyJpIjoiYTNlZDUyMzllYzc3NDU0YjgxYTUxNjdiNjc5Y2I5YTYiLCJwIjoiaiJ9

File: config/reader.go
Comment (zaporter-work) [inline config/reader.go:572]: [nil] This should come after checking the err from `service.Config` otherwise res could be nil / empty and result in a segfault

**Review by zaporter-work — APPROVED** [review]
LGTM

File: module/modmanager/manager.go
Comment (zaporter-work) [inline module/modmanager/manager.go:1]: Watch out for https://github.com/viamrobotics/rdk/pull/3095

I think this will have the same bug.

**Review by zaporter-work — CHANGES_REQUESTED** [review]

File: module/modmanager/manager.go
Comment (michaellee1019) [reply module/modmanager/manager.go:1]: Good point. I think it will avoid the issue for new RDK versions, since the fields are added to ModuleConfigFromProto and PackageConfigFromProto in advance. It does mean:

1. Need to release RDK first, then App.
2. Need to have app pass a nil status field for older RDK version.

I think this is correct?

**Review by michaellee1019 — APPROVED** [review]

File: module/modmanager/manager.go
Comment (zaporter-work) [reply module/modmanager/manager.go:1]: I decided to fix this footgun in https://github.com/viamrobotics/rdk/pull/3178

You no longer need to modify modmanager or anything like that. Now just need to release RDK and then bring this into app.

**Comment by RoxyFarhad** [comment]
> Need to release RDK first, then App.
Need to have app pass a nil status field for older RDK version.

Yes that is correct! We want to merge the change to handle the errors in RDK before we merge in the APP change. So we will also need to wait for RDK new version to be released, before we also release APP.

File: config/config.go
Comment (zaporter-work) [inline config/config.go:85]:
```suggestion
	Error string `json:"error"`
```

The rdk configs go back and forth in JSON, not bson. See:

```

	// Status refers to the validations done in the APP to make sure a module is configured correctly
	Status           *appValidationStatus `json:"status"`
```

File: config/module.go
Comment (zaporter-work) [inline config/module.go:39]: I am slightly worried that this is a pointer. None of our other config objects contain pointers (that aren't populated inside the rdk) (that I know of). It is *probably* fine...

The reason I am worried is that it means marshalling and unmarshalling this struct will result in a different pointer, which will lead to differences. (I'm guessing this is why you set status to nil in `Equals`). 

Would it be okay to make this a non-pointer type? Then I don't think you need the `nil` sets in `Equals()`

Up to you.

**Review by zaporter-work — COMMENTED** [review]
Would you mind updating the to/from proto tests to include an example of you converting this new type? 

I think that would be a good thing to have. However, if you'd rather break that into a diff ticket, probably okay.

File: config/config.go
Comment (RoxyFarhad) [reply config/config.go:954]: ah didn't see that a nice wrapper! adding

File: config/proto_conversions.go
Comment (RoxyFarhad) [reply config/proto_conversions.go:266]: yeah i mean both doing the same thing, i can change this so it looks the same but in theory there isn't an output diff if either are set/nil etc.

File: config/config.go
Comment (RoxyFarhad) [inline config/config.go:958]: @benjirewis just wanted to check something with you on this PR. Should we be storing this as a cached error? This new error is essentially a validation error from app to show that something is wrong with the module.

File: config/module.go
Comment (RoxyFarhad) [reply config/module.go]: Yeah, it makes sense we don't do it for primitive data types because they will just be set to an empty string which is easy to check but for nested objects its a bit harder.

**Review by zaporter-work — APPROVED** [review]
LGTM!

File: config/config.go
Comment (benjirewis) [reply config/config.go:958]: I guess I would yeah. `Validate`ion usually happens multiple times for a single package during a reloading of the config/reconfiguration of the robot. Eric put in this caching system to stop running the full `Validate` multiple times for every resource/module/package. I would `p.alreadyValidated = true; p.cachedErr = utils.NewConfigValidationError(...); return p.cachedErr` here or something of the like. It's not the end of the world if you don't, but you'll be checking that `p.Status` field and creating a new error multiple times unnecessarily.

**Comment by RoxyFarhad** [comment]
Just to sanity check too, running the current version of APP with this version of RDK does not break RDK and maintains the same behaviour. config will return nil and it will check the cache for the last updated version.

**Review by michaellee1019 — APPROVED** [review]

File: config/config.go
Comment (RoxyFarhad) [reply config/config.go]: okay sounds good I changed based on the convo ^

**Pull request merged**
